vm-virtio/src/net.rs

```diff
@@ -626,24 +516,46 @@ impl VirtioDevice for Net {
             }
             self.queue_evts = Some(tmp_queue_evts);
 
-            let rx_queue = queues.remove(0);
-            let tx_queue = queues.remove(0);
-            let rx_queue_evt = queue_evts.remove(0);
-            let tx_queue_evt = queue_evts.remove(0);
+            let queue_num = queues.len();
+
+            let (taps_v, rxs, txs, rx_tap_listenings) =
+                build_net_element_collector(taps, queue_num);
+
+            let ctrl_q = if (self.acked_features & 1 << VIRTIO_NET_F_CTRL_VQ) != 0 {
+                if queues.len() % 2 != 0 {
+                    let cvq_queue = queues.pop().unwrap();
+                    let cvq_queue_evt = queue_evts.pop().unwrap();
+                    Some(CtrlVirtio::new(cvq_queue, cvq_queue_evt))
+                } else {
+                    None
+                }
+            } else {
+                None
+            };
+
             let mut handler = NetEpollHandler {
                 mem,
-                tap,
-                rx: RxVirtio::new(rx_queue, rx_queue_evt),
-                tx: TxVirtio::new(tx_queue, tx_queue_evt),
+                taps: taps_v,
+                rxs,
+                txs,
                 interrupt_cb,
                 kill_evt,
                 epoll_fd: 0,
-                rx_tap_listening: false,
+                rx_tap_listenings,
+                ctrl_q,
             };
 
-            let worker_result = thread::Builder::new()
-                .name("virtio_net".to_string())
-                .spawn(move || handler.run());
+            let worker_result =
+                thread::Builder::new()
```

Comment:
You need to start a thread per queue to actually get any benefit from MQ.

Reply:
Yes, I start multi-thread in local repo for test.